pep-perennial.rst

```diff
@@ -0,0 +1,161 @@
+PEP: XXX
+Title: Future 'manylinux' Platform Tags for Portable Linux Built Distributions
+Version: $Revision$
+Last-Modified: $Date$
+Author: Nathaniel J. Smith <njs@pobox.com>
+        Thomas Kluyver <thomas@kluyver.me.uk>
+BDFL-Delegate: Nick Coghlan <ncoghlan@gmail.com>
+Discussions-To: Distutils SIG <distutils-sig@python.org>
+Status: Active
+Type: Informational
+Content-Type: text/x-rst
+Created: 3-May-2019
+Post-History: 3-May-2019
+
+Abstract
+========
+
+This PEP proposes a scheme for new 'manylinux' distribution tags to be defined
+without requiring a PEP for every specific tag. The naming scheme is based on
+glibc versions, with profiles in the auditwheel tool defining what other
+external libraries and symbols a compatible wheel may link against.
+
+While there is interest in defining tags for non-glibc Linux platforms,
+this PEP does not attempt to address that.
+
+Rationale
+=========
+
+Distributing compiled code for Linux is more complicated than for other popular
+operating systems, because the Linux kernel and the libraries which typically
+accompany it are built in different configurations and combinations for different
+distributions. However, there are certain core libraries which can be expected in
+many common distributions and are reliably backwards compatible, so binaries
+built with older versions of these libraries will work with newer versions.
+:pep:`513` describes these ideas in much more detail.
+
+The ``manylinux1`` (:pep:`513`) and ``manylinux2010`` (:pep:`571`) tags make
+use of these features. They define a set of core libraries and symbol versions
+which wheels may expect on the system, based on CentOS 5 and 6 respectively.
+Typically, packages are built in Docker containers based on these old CentOS
+versions, and then the ``auditwheel`` tool is used to check them and bundle any
+other linked libraries into the wheel.
+
+If we were to define a ``manylinux2014`` tag based on CentOS 7, there would be
+five steps involved to make it practically useful:
+
+1. Write a PEP
+2. Prepare docker images based on CentOS 7.
+3. Add the definition to auditwheel
+4. Allow uploads with the new tag on PyPI
+5. Add code to pip to recognise the new tag and check if the platform is
+   compatible
+
+Although preparing the docker images and updating auditwheel take more work,
+these parts can be used as soon as that work is complete. The changes to pip
+are more straightforward, but not all users will promptly install a new version
+of pip, so distributors are concerned about moving to a new tag too quickly.
+
+This PEP aims to remove the need for steps 1 and 5 above, so new manylinux tags
+can be adopted more easily.
```

Comment:
My assumption is that Warehouse would need to bump the version of auditwheel it's using to pick up the new profile, so there's still a small change to make.

Reply:
Eh, that's kind of orthogonal to this discussion (and anyway they have a bot that handles version bumping), so I'd probably count it? It would remove the need to explicitly write code and write a test. But it's not a big deal either way.